secret_store/src/key_server_cluster/cluster.rs

```diff
@@ -803,12 +804,28 @@ impl ClusterData {
 		self.connections.get(node)
 	}
 
-	/// Spawns a future using thread pool and schedules execution of it with event loop handle.
-	pub fn spawn<F>(&self, f: F) where F: Future + Send + 'static, F::Item: Send + 'static, F::Error: Send + 'static {
-		let pool_work = self.pool.spawn(f);
-		self.handle.spawn(move |_handle| {
-			pool_work.then(|_| finished(()))
-		})
+	/// Spawns a future on the runtime.
+	//
+	// TODO: Consider implementing a more graceful shutdown process using an
+	// `AtomicBool`, etc. which would prevent tasks from being spawned after a
+	// shutdown signal is given. (Recursive calls, in
+	// `process_connection_messages` for example, appear to continue
+	// indefinitely.)
+	pub fn spawn<F>(&self, f: F) where F: Future<Item = (), Error = ()> + Send + 'static {
+		if self.is_shutdown.load(Ordering::Acquire) == false {
+			if let Err(err) = future::Executor::execute(&self.executor, Box::new(f)) {
+				error!("Secret store runtime unable to spawn task. Runtime is shutting down. ({:?})", err);
+			}
+		} else {
+			error!("Secret store runtime unable to spawn task. Shutdown has been started.");
+		}
+	}
+
+	/// Sets the `is_shutdown` flag which prevents future tasks from being
+	/// spawned via `::spawn`.
+	#[cfg(test)]
+	pub fn shutdown(&self) {
+		self.is_shutdown.store(true, Ordering::Release);
```

Comment:
What's the reason for this method? Something bad happened in tests?

Reply:
Basically, what was happening without that is that due to the recursive nature of task spawning, one task could be spawning new tasks as the runtime is shutting down, causing errors. Those errors sometimes caused tests to fail but usually didn't.

This probably shouldn't be just a test only thing but let's leave that for later.